src/mbgl/text/placement.cpp

```diff
@@ -186,162 +184,175 @@ void Placement::placeLayerBucket(
     std::vector<style::TextVariableAnchorType> variableTextAnchors = bucket.layout.get<style::TextVariableAnchor>();
     const bool rotateWithMap = bucket.layout.get<style::TextRotationAlignment>() == style::AlignmentType::Map;
     const bool pitchWithMap = bucket.layout.get<style::TextPitchAlignment>() == style::AlignmentType::Map;
-    
-    for (SymbolInstance& symbolInstance : bucket.symbolInstances) {
 
-        if (seenCrossTileIDs.count(symbolInstance.crossTileID) == 0) {
-            if (holdingForFade) {
-                // Mark all symbols from this tile as "not placed", but don't add to seenCrossTileIDs, because we don't
-                // know yet if we have a duplicate in a parent tile that _should_ be placed.
-                placements.emplace(symbolInstance.crossTileID, JointPlacement(false, false, false));
-                continue;
-            }
+    const bool zOrderByViewportY = bucket.layout.get<style::SymbolZOrder>() == style::SymbolZOrderType::ViewportY;
```

Comment:
`!= style::SymbolZOrderType::Source`

Reply:
That will affect the default behavior for styles that did not specify `symbol-z-order` :thinking:

Reply:
I think we want to avoid sorting for all layers by default since this will introduce more label flickering. I think your implemented approach is good.

One alternative would be the same logic used to decide whether to sort the symbols when rendering: https://github.com/mapbox/mapbox-gl-js/blob/29a27eba5ff79720fa184ec4dc8b9dd942f759f9/src/data/bucket/symbol_bucket.js#L318-L319 - the symbols are only sorting if they are allowed to overlap. This same kind of logic could be used here but I'm not sure if that is necessary or better.

I think I like your approach better because it is simpler.

Reply:
IMO checking just `SymbolZOrder` is fine for placement: it just tells which labels to be placed first, so it is functional even if both icon and text overlapping not allowed (unlike the situation in bucket).